Support for tcp transport

M-Bab opened the pull request:
Implementation for D-Bus communication via TCP/IP. Previously there was also AUTH ANONYMOUS added to the authentifications but this is now already implemented - so no need to change anything there.

acrisci commented:
Is it possible to write tests for this?

M-Bab commented:
It is - if there is a CI system available where a tcp-dbus connection is possible. Currently we are using a special test system where the dbus is open to the network and allows anonymous authentification (very evil).
   
  
  
   
  
  
   I will try out if a simple, safe and effective test can be implemented with socat on the loopback interface.
   
  
  
   
  
  
   Tony Crisci <notifications@github.com> hat am 23. August 2020 um 21:19 geschrieben: 
   
   
   
   Is it possible to write tests for this?
   —You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
  
  
    

M-Bab commented:
Okay I managed to get it fully working for a local test:

1. Open a local socket:
`socat TCP-LISTEN:55556,reuseaddr,fork,range=127.0.0.1/32 UNIX-CONNECT:$(echo $DBUS_SESSION_BUS_ADDRESS | sed 's/unix:path=//g')`

2. This piece of code creates a notification via tcp:
```python
import asyncio
import sys
from dbus_next.aio import MessageBus
#from dbus_next.message import Message, MessageType

async def main(args):
  bus = await MessageBus(bus_address="tcp:host=127.0.0.1,port=55556").connect()
  #reply = await bus.call(Message(destination='org.freedesktop.DBus',path='/org/freedesktop/DBus',interface='org.freedesktop.DBus',member='ListNames'))
  introspection = await bus.introspect('org.freedesktop.Notifications', '/org/freedesktop/Notifications')
  obj = bus.get_proxy_object('org.freedesktop.Notifications', '/org/freedesktop/Notifications', introspection)
  notification = obj.get_interface('org.freedesktop.Notifications')
  await notification.call_notify("test.py", 0, "", "DBus Test", "Test notification", [""], dict(), 3000)

if __name__ == "__main__":
  loop = asyncio.get_event_loop()
  basefilename = loop.run_until_complete(main(sys.argv))
```
It is somehow amazing that it works because neither d-feet nor notify-send were able to connect to the socat in my case.

Or what kind of tests specifically shall I add?

acrisci commented:
The only way I can ensure this will be stable with future changes if you add tests to the [test suite](https://github.com/altdesktop/python-dbus-next/tree/master/test) for your use cases.

M-Bab commented:
Here you go :+1:. Added 2 tests which are simple but should be very effective to protect the functionality. Also added an example for anyone willing to use it.

acrisci commented on test/test_aio_low_level.py:
```diff
@@ -136,3 +137,14 @@ def message_handler(signal):
+@pytest.mark.asyncio
+async def test_connection_via_tcp():
+    server = await asyncio.start_server(None, '127.0.0.1', 55556)
+    bus = MessageBus(bus_address="tcp:host=127.0.0.1,port=55556")
```
Can you call `connect()` to send a message across?

M-Bab replied:
Basically yes. You can specify the callback function and receive the "AUTH EXTERNAL ..." which can be replied with "OK" but then the system got stock (probably due to a lack of the Hello message).

But thats not ultimately necessary to unit test this function because the proof of a network connection seams good enough for me.

acrisci replied:
I'd really like to see a message sent across the connection and successful auth annonymous. #54 for example will add assumptions that the socket is unix that would break this feature and have these tests still pass.

For an example of how to add dbus configuration to the test suites to listen on a tcp address, see the [playerctl dockerfile](https://github.com/altdesktop/playerctl/blob/master/Dockerfile) and test suite which is based on this project.

M-Bab replied:
I see the point and I tried to get it working locally - but it did not work before the weekend ended. Somehow the session.conf line order fragility, systemd and even apparmor joined forces to prevent any feeling of success :disappointed: 

Maybe I am also not clever enough or at least not experienced enough in DBus/Docker for this. If you don't want to merge it, it's not a problem either because my company can live with my fork as well. The possibility of a TCP connection is all they needed - I just thought it would be nice to upstream it for other users and to ensure a future for this feature.

acrisci replied:
Ok then this might be a bit more complicated. I can merge it with an issue to create tests if it's a big task. How exactly are you using it? The usage example you gave did some kind of socket forwarding.

acrisci commented:
This is fine for now :+1:

M-Bab commented:
Thanks a lot. I am sorry I could not get a good working test for the combination network+anonymous auth via dbus config. Some colleagues of mine figured it out on an embedded device but it didn't work for me on any x86 machine.
